### PR TITLE
Add Resize Handler for Transpose Optimizer and revise Layout Transformation API

### PR DESCRIPTION
**Pull request by TomWildenhain-Microsoft**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/optimizer/transpose_optimizer/api.h [hunk 1]
```diff
@@ -72,19 +72,35 @@ class TensorRef {
   /// <returns>The dtype of the tensor.</returns>
   virtual DataType DType() const = 0;
 
+  /// <summary>
+  /// Retrieves raw data from the tensor. Used for reading initializers specifying axes/pads/scales.
+  /// </summary>
+  /// <returns>Flattened tensor data in bytes</returns>
+  virtual std::vector<char> Data() const = 0;
+
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/api.h:80]: What's the benefit of this to justify the reinterpret_cast's and multiple copies of the data (e.g. Tensor -> vector<char> -> vector<intX>)?

Is it mainly so PermuteInput doesn't have a branch on int64 vs int32?

File: onnxruntime/core/optimizer/transpose_optimizer/api.h [hunk 2]
```diff
@@ -430,60 +473,72 @@ constexpr int64_t kMaxSupportedOpset = 15;
...
- * ChannelFirstToChannelLast. If the parameters of these convenience methods are too restrictive, you can swap the
- * ops for reordered versions yourself and call Optimize manually.
+ * These methods help change the channel ordering of layout sensitive ops (like Conv). ONNX currently only supports
+ * channel first ordering for ops, so this requires changing the op type and domain to a contrib op supporting
+ * the new ordering. The existence of a robust transpose optimizer means that we can freely add transpose ops during
+ * conversion and then call Optimize to remove as many as possible. To change the channel ordering of some/all ops
+ * in a model, a user of this tool should do the following:
+ * 
+ * 1. Iterate over the graph nodes and identify nodes for convert. For each one:
+ *    a. Change the op type and domain (and possibly attributes) to the op/contrib op with the desired ordering.
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/api.h:483]: nit: nodes to convert

File: onnxruntime/core/optimizer/transpose_optimizer/api.h [hunk 3]
```diff
@@ -430,60 +473,72 @@ constexpr int64_t kMaxSupportedOpset = 15;
...
-/// will persist.
+/// Inserts transposes around op inputs/outputs. Alternatively transposes initializers or uses existing Transpose
+/// nodes if possible. Populates shape information on affected node inputs/outputs to reflect the change.
+/// 
+/// Ex:
+///   * -> Conv -> **
+///   becomes
+///   * -> Transpose -> Conv -> Transpose -> **
+///   Model is now invalid until Conv op type is changed. Conv inputs/outputs have new shape. Shapes of * and **
+///   are unchanged.
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/api.h:502]: nit: this doesn't seem to match with above where it says step 1 is to change the op type and domain (rendering the graph invalid) and step 2 is to insert the Transposes to fix. 

would have expected a line in the middle with * --> Conv.NHWC -> ** or something like that

File: onnxruntime/core/optimizer/nhwc_transformer.cc [hunk 4]
```diff
@@ -14,48 +14,56 @@ using namespace onnx_layout_transformation;
...
+      modified = true;
+    }
+  }
 
-  auto api_graph = MakeApiGraph(graph, std::move(cpu_allocator_), logger, kCpuExecutionProvider);
-  if (ChannelFirstToChannelLast(*api_graph, handler_map, /*allow_extended_ops*/ true)) {
-    modified = true;
+  if (modified) {
+    Optimize(*api_graph, /*allow_extended_ops*/ true);
   }
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/nhwc_transformer.cc:67]: If Optimize recurses does it only need to be called when level is 1? 

We do that with Graph::Resolve, but it's slightly different in that it must always start at the main graph even if called from a subgraph.

Comment (TomWildenhain-Microsoft) [reply onnxruntime/core/optimizer/nhwc_transformer.cc:67]: Optimize only optimizes the current graph level.

File: onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc [hunk 5]
```diff
@@ -48,6 +48,34 @@ struct HandlerInfo {
 /////// <Helper Utils> ///////
 /* Small utilities for editing nodes and manipulating axes/permutations */
 
+std::vector<int64_t> DataInt64(api::TensorRef& tensor) {
+  std::vector<uint8_t> raw_data = tensor.Data();
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc:52]: nit: can these all be static?

File: onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc [hunk 6]
```diff
@@ -426,40 +458,40 @@ static void UnsqueezeInput(OptimizerCtx& ctx, api::NodeRef& node, size_t i, cons
...
-static void TransposeInput(OptimizerCtx& ctx, api::NodeRef& node, size_t i,
+static void TransposeInput(api::GraphRef& graph, api::NodeRef& node, size_t i,
                            const std::vector<int64_t>& perm, const std::vector<int64_t>& perm_inv) {
   std::string_view input = node.Inputs()[i];
   // Remove this node as a consumer
   node.SetInput(i, "");
-  std::unique_ptr<api::TensorRef> constant = ctx.graph.GetConstant(input);
-  auto consumers = ctx.graph.GetValueConsumers(input);
+  std::unique_ptr<api::TensorRef> constant = graph.GetLocalConstant(input);
+  auto consumers = graph.GetValueConsumers(input);
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc:467]: maybe add a comment explaining why we use GetLocalConstant

File: onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc [hunk 7]
```diff
@@ -866,73 +899,111 @@ static bool HandleShape(HandlerArgs& args) {
 
 constexpr HandlerInfo shape_handler = {&FirstInput, &HandleShape, /*transposes_outputs*/ false};
 
-// Reorder pads according to perm. Pads length is twice perm length (all starts then all ends).
-static std::vector<int64_t> PermutePads(const std::vector<int64_t>& pads, const std::vector<int64_t>& perm) {
+// Permutes a 1D node input by creating a new initializer or inserting a Gather op
+void PermuteInput(api::GraphRef& graph, api::NodeRef& node, size_t i, std::vector<int64_t> perm) {
   size_t rank = perm.size();
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc:904]: nit can `perm` be a const reference?

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]

**Pull request merged**